// @bedrockio/yada/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LocalizedError", {
  enumerable: true,
  get: function () {
    return _errors.LocalizedError;
  }
});
exports.allow = allow;
exports.any = any;
Object.defineProperty(exports, "array", {
  enumerable: true,
  get: function () {
    return _array.default;
  }
});
Object.defineProperty(exports, "boolean", {
  enumerable: true,
  get: function () {
    return _boolean.default;
  }
});
exports.custom = custom;
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function () {
    return _date.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "getLocalizedMessages", {
  enumerable: true,
  get: function () {
    return _localization.getLocalizedMessages;
  }
});
Object.defineProperty(exports, "isSchema", {
  enumerable: true,
  get: function () {
    return _utils.isSchema;
  }
});
Object.defineProperty(exports, "isSchemaError", {
  enumerable: true,
  get: function () {
    return _utils.isSchemaError;
  }
});
Object.defineProperty(exports, "number", {
  enumerable: true,
  get: function () {
    return _number.default;
  }
});
Object.defineProperty(exports, "object", {
  enumerable: true,
  get: function () {
    return _object.default;
  }
});
exports.reject = reject;
Object.defineProperty(exports, "string", {
  enumerable: true,
  get: function () {
    return _string.default;
  }
});
Object.defineProperty(exports, "tuple", {
  enumerable: true,
  get: function () {
    return _tuple.default;
  }
});
Object.defineProperty(exports, "useLocalizer", {
  enumerable: true,
  get: function () {
    return _localization.useLocalizer;
  }
});
var _string = _interopRequireDefault(require("./string"));
var _number = _interopRequireDefault(require("./number"));
var _boolean = _interopRequireDefault(require("./boolean"));
var _date = _interopRequireDefault(require("./date"));
var _object = _interopRequireDefault(require("./object"));
var _array = _interopRequireDefault(require("./array"));
var _tuple = _interopRequireDefault(require("./tuple"));
var _Schema = _interopRequireDefault(require("./Schema"));
var _utils = require("./utils");
var _localization = require("./localization");
var _errors = require("./errors");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/**
 * Accepts anything.
 */
function any() {
  return new _Schema.default();
}

/**
 * Accept values or schemas. [Link](https://github.com/bedrockio/yada#allow)
 */
function allow(...args) {
  return new _Schema.default().allow(...args);
}

/**
 * Reject values or schemas. [Link](https://github.com/bedrockio/yada#reject)
 */
function reject(...args) {
  return new _Schema.default().reject(...args);
}

/**
 * Validate by a custom function. [Link](https://github.com/bedrockio/yada#custom)
 * @param {import("./Schema").CustomSignature} args
 */
function custom(...args) {
  return new _Schema.default().custom(...args);
}
var _default = {
  array: _array.default,
  boolean: _boolean.default,
  date: _date.default,
  number: _number.default,
  object: _object.default,
  string: _string.default,
  tuple: _tuple.default,
  any,
  allow,
  reject,
  custom,
  isSchema: _utils.isSchema,
  isSchemaError: _utils.isSchemaError,
  useLocalizer: _localization.useLocalizer,
  getLocalizedMessages: _localization.getLocalizedMessages,
  LocalizedError: _errors.LocalizedError
};
exports.default = _default;
