// @bedrockio/logger/dist/cjs/loggers/FormattedLogger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _kleur = require("kleur");
var _ConsoleLogger = _interopRequireDefault(require("./ConsoleLogger"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
class FormattedLogger extends _ConsoleLogger.default {
  print(level, ...args) {
    const fn = console[level];
    let msg = `${this.getDateTag()} ${this.getLevelTag(level)}`;
    if (typeof args[0] === 'string') {
      const [first, ...rest] = args;
      msg += ` ${first}`;
      args = rest;
    }
    fn(msg, ...args);
  }
  getDateTag() {
    // Local date in ISO format, no ms.
    const date = new Date();
    date.setMinutes(date.getMinutes() - date.getTimezoneOffset());
    return (0, _kleur.gray)(`[${date.toISOString().slice(0, -5)}]`);
  }
  getLevelTag(level) {
    let tag = level.toUpperCase().padStart(5, ' ');
    return this.formatForLevel(level, tag);
  }
  formatStatus(status) {
    if (status >= 500) {
      return (0, _kleur.red)(status);
    } else if (status >= 400) {
      return (0, _kleur.yellow)(status);
    } else if (status >= 300) {
      return (0, _kleur.cyan)(status);
    } else {
      return (0, _kleur.green)(status);
    }
  }
  formatMeta(meta) {
    return (0, _kleur.gray)(meta);
  }
}
exports.default = FormattedLogger;
module.exports = exports.default;
