// @bedrockio/logger/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "debug", {
  enumerable: true,
  get: function () {
    return _logger.debug;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "error", {
  enumerable: true,
  get: function () {
    return _logger.error;
  }
});
Object.defineProperty(exports, "formatRequest", {
  enumerable: true,
  get: function () {
    return _logger.formatRequest;
  }
});
Object.defineProperty(exports, "info", {
  enumerable: true,
  get: function () {
    return _logger.info;
  }
});
Object.defineProperty(exports, "middleware", {
  enumerable: true,
  get: function () {
    return _middleware.default;
  }
});
exports.setupGoogleCloud = setupGoogleCloud;
Object.defineProperty(exports, "trace", {
  enumerable: true,
  get: function () {
    return _logger.trace;
  }
});
Object.defineProperty(exports, "useConsole", {
  enumerable: true,
  get: function () {
    return _logger.useConsole;
  }
});
Object.defineProperty(exports, "useFormatted", {
  enumerable: true,
  get: function () {
    return _logger.useFormatted;
  }
});
Object.defineProperty(exports, "useGoogleCloud", {
  enumerable: true,
  get: function () {
    return _logger.useGoogleCloud;
  }
});
Object.defineProperty(exports, "warn", {
  enumerable: true,
  get: function () {
    return _logger.warn;
  }
});
var _logger = require("./logger");
var _middleware = _interopRequireDefault(require("./middleware"));
var _tracing = require("./tracing");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const DEFAULT_OPTIONS = {
  logging: true,
  tracing: true
};

/**
 * @param {Object} [options]
 * @param {boolean} [options.logging=true]
 * @param {boolean|Object} [options.tracing=true]
 */
function setupGoogleCloud(options) {
  options = {
    ...DEFAULT_OPTIONS,
    ...options
  };
  if (options.logging) {
    (0, _logger.useGoogleCloud)({
      getTracePayload: _tracing.getTracePayload
    });
  }
  if (options.tracing) {
    (0, _tracing.useGoogleCloudTracing)({
      ignoreIncomingPaths: options.tracing?.ignoreIncomingPaths
    });
  }
}
var _default = {
  trace: _logger.trace,
  debug: _logger.debug,
  info: _logger.info,
  warn: _logger.warn,
  error: _logger.error,
  setupGoogleCloud,
  formatRequest: _logger.formatRequest,
  middleware: _middleware.default,
  useConsole: _logger.useConsole,
  useFormatted: _logger.useFormatted,
  useGoogleCloud: _logger.useGoogleCloud
};
exports.default = _default;
