// @bedrockio/logger/dist/cjs/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.debug = debug;
exports.error = error;
exports.formatRequest = formatRequest;
exports.info = info;
exports.trace = trace;
exports.useConsole = useConsole;
exports.useFormatted = useFormatted;
exports.useGoogleCloud = useGoogleCloud;
exports.warn = warn;
var _ConsoleLogger = _interopRequireDefault(require("./loggers/ConsoleLogger"));
var _FormattedLogger = _interopRequireDefault(require("./loggers/FormattedLogger"));
var _GoogleCloudLogger = _interopRequireDefault(require("./loggers/GoogleCloudLogger"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
let logger;
function useGoogleCloud(options) {
  logger = new _GoogleCloudLogger.default(options);
}
function useFormatted() {
  logger = new _FormattedLogger.default();
}
function useConsole() {
  logger = new _ConsoleLogger.default();
}
function trace(...args) {
  return logger['trace'](...args);
}
function debug(...args) {
  return logger['debug'](...args);
}
function info(...args) {
  return logger['info'](...args);
}
function warn(...args) {
  return logger['warn'](...args);
}
function error(...args) {
  return logger['error'](...args);
}

/**
 * @param {Object} request
 * @param {string} request.method
 * @param {string} request.path
 * @param {number} request.status
 * @param {number} request.latency
 * @param {string} request.size
 *
 * Formats a log for a request object from Koa.
 */
function formatRequest(request) {
  return logger['formatRequest'](request);
}

// Default to console logger
useConsole();
